File: front/mini-app/src/api/changeEgg.ts
import { request } from '@/utils/request'

export type ChangeEggStatus = 'open' | 'matched' | 'closed'

export interface ChangeEggListing {
  id: number
  user_id: number
  game_id: string
  own_pokemon_id: number
  own_tag: string
  want_pokemon_id: number
  want_tag: string
  status: ChangeEggStatus | string
  own_pokemon_name: string | null
  own_pokemon_avatar: string | null
  want_pokemon_name: string | null
  want_pokemon_avatar: string | null
  created_at: string | null
  updated_at: string | null
}

export interface ChangeEggCreatePayload {
  user_id: number | string
  game_id: string
  own_pokemon_id: number
  own_tag: string
  want_pokemon_id: number
  want_tag: string
}

export interface ChangeEggTag {
  code: string
  label: string
  sort_order: number
}

/** 查询我发布的换蛋挂单，可选按状态筛选。 */
export function fetchMyListings(userId: number | string, status?: ChangeEggStatus | '') {
  return request<ChangeEggListing[]>({
    url: '/api/change-eggs',
    data: { user_id: userId, status: status || undefined },
  })
}

/** 广场分页浏览所有匹配中的挂单。 */
export function fetchSquareListings(params: {
  limit?: number
  offset?: number
  pokemon_id?: number
  tag?: string
} = {}) {
  return request<ChangeEggListing[]>({
    url: '/api/change-eggs/square',
    data: params,
  })
}

/** 可选的蛋组 tag（大块头 / 小不点等）。 */
export function fetchEggTags() {
  return request<ChangeEggTag[]>({ url: '/api/change-eggs/tags' })
}

/** 发布一条换蛋挂单。 */
export function createListing(payload: ChangeEggCreatePayload) {
  return request<ChangeEggListing>({
    url: '/api/change-eggs',
    method: 'POST',
    data: payload,
  })
}

/** 主动关闭正在匹配（open）的挂单。 */
export function closeListing(listingId: number, userId: number | string) {
  return request<ChangeEggListing>({
    url: `/api/change-eggs/${listingId}/close`,
    method: 'POST',
    data: { user_id: userId },
  })
}

/** 彻底删除自己的挂单（user_id 走 query，便于后端 Query 校验归属）。 */
export function deleteListing(listingId: number, userId: number | string) {
  return request<{ ok: boolean }>({
    url: `/api/change-eggs/${listingId}?user_id=${encodeURIComponent(String(userId))}`,
    method: 'DELETE',
  })
}

File: api/schemas/change_egg.py
from datetime import datetime

from pydantic import BaseModel, Field


class ChangeEggCreateRequest(BaseModel):
    """新增换蛋挂单：拥有(pokemon_id+tag) + 需求(pokemon_id+tag) + 游戏 id。"""

    user_id: int = Field(..., description="平台用户 id")
    game_id: str = Field("", max_length=64, description="游戏 id（长串数字）")
    own_pokemon_id: int = Field(..., description="拥有的精灵 id（pokemon.id）")
    own_tag: str = Field("", max_length=30, description="拥有蛋组的 tag code")
    want_pokemon_id: int = Field(..., description="需求的精灵 id（pokemon.id）")
    want_tag: str = Field("", max_length=30, description="需求蛋组的 tag code")


class ChangeEggListing(BaseModel):
    """一条换蛋挂单（含 own/want 对应 pokemon 的展示信息）。"""

    id: int
    user_id: int
    game_id: str = ""
    own_pokemon_id: int
    own_tag: str = ""
    want_pokemon_id: int
    want_tag: str = ""
    status: str = "open"
    own_pokemon_name: str | None = None
    own_pokemon_avatar: str | None = None
    want_pokemon_name: str | None = None
    want_pokemon_avatar: str | None = None
    created_at: datetime | None = None
    updated_at: datetime | None = None


class ChangeEggCloseRequest(BaseModel):
    """主动关闭自己正在匹配的挂单。"""

    user_id: int = Field(..., description="平台用户 id（校验归属）")


class ChangeEggTag(BaseModel):
    """换蛋蛋组 tag 选项。"""

    code: str
    label: str
    sort_order: int = 0
